File: src/controller/DB/Tables/ChannelTable.ts
interface ChannelTable {
	name: string;
	user_id: string;
	/**
	 * Bot permission defines the permission level of the bot in the channel.
	 *
	 * {0 = Read} Means the bot won't respond to any messages.
	 *
	 * {1 = Write} Bot behaves normally, allowing all commands.
	 *  Message interval -- 1.25 seconds.
	 *
	 * {2 = VIP}
	 *  Message interval -- 0.25 seconds.
	 *
	 * {3 = Moderator}
	 *  Message interval -- 0.05 seconds.
	 *
	 * {4 = Bot} Own channel
	 */
	bot_permission: number;
}

type PermissionMode = 'Read' | 'Write' | 'VIP' | 'Moderator' | 'Bot';

/**
 * Converts the database stored permission value to a mode.
 */
function ChannelDatabaseToMode(value: number): PermissionMode {
	switch (value) {
		case 0:
			return 'Read';
		case 1:
			return 'Write';
		case 2:
			return 'VIP';
		case 3:
			return 'Moderator';
		case 4:
			return 'Bot';
		default:
			return 'Read';
	}
}

function PermissionModeToDatabase(mode: PermissionMode): number {
	switch (mode) {
		case 'Read':
			return 0;
		case 'Write':
			return 1;
		case 'VIP':
			return 2;
		case 'Moderator':
			return 3;
		case 'Bot':
			return 4;
		default:
			return 0;
	}
}

/**
 * Converts a permission mode to the equivalent message-interval value.
 */
function PermissionModeToCooldown(mode: PermissionMode): number | null {
	switch (mode) {
		case 'Read':
			return null;
		case 'Write':
			return 1250;
		case 'VIP':
			return 250;
		case 'Moderator':
			return 50;
		case 'Bot':
			return 0;
		default:
			return null;
	}
}

export default ChannelTable;
export {
	ChannelTable,
	PermissionMode,
	PermissionModeToCooldown,
	ChannelDatabaseToMode,
	PermissionModeToDatabase,
};

File: Golang/internal/models/dbmodels/channel_table.go
package dbmodels

type BotPermmision int

func (b BotPermmision) String() string {
	switch b {
	case ReadPermission:
		return "Read"
	case WritePermission:
		return "Write"
	case VIPPermission:
		return "VIP"
	case ModeratorPermission:
		return "Moderator"
	case BotPermission:
		return "Bot"
	default:
		return "Unknown"
	}
}

// Returns the message interval cooldown in milliseconds a bot is allowed to send messages to a channel
func (b BotPermmision) ToMessageCooldown() int {
	switch b {
	case ReadPermission, WritePermission:
		return DEFAULT_MESSAGE_INTERVAL
	case VIPPermission:
		return 250
	case ModeratorPermission, BotPermission:
		return 50
	default:
		return DEFAULT_MESSAGE_INTERVAL
	}
}

const (
	// The bot is not allowed to write nor react to messages in the channel
	ReadPermission BotPermmision = 0
	// Default permission levels
	WritePermission BotPermmision = 1
	// Bot is a VIP in the channel
	VIPPermission BotPermmision = 2
	// Bot is a moderator in the channel
	ModeratorPermission BotPermmision = 3
	// The bot's own channel
	BotPermission BotPermmision = 4

	// Default message interval in milliseconds
	DEFAULT_MESSAGE_INTERVAL = 1250
)

type ChannelTable struct {
	Name          string `gorm:"column:name"`
	UserID        string `gorm:"column:user_id"`
	BotPermission int    `gorm:"column:bot_permission"`
}

func (ChannelTable) TableName() string {
	return "bot.channels"
}

func (c *ChannelTable) GetBotPermission() BotPermmision {
	return BotPermmision(c.BotPermission)
}

// Determines if a bot is allowed to speak based on the bot permission level
func (c *ChannelTable) IsAllowedToSpeak() bool {
	return c.GetBotPermission() != ReadPermission
}
